### PR TITLE
Ensure multiple post types don’t throw a notice.

### PR DESCRIPTION
**Pull request by peterwilsoncc**
Trac ticket: https://core.trac.wordpress.org/ticket/43056

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: tests/phpunit/tests/canonical.php [hunk 1]
```diff
@@ -275,6 +275,19 @@ public function test_strict_redirect_guess_404_permalink() {
...
 	}
 
+	/**
+	 * Ensure multiple post types do not throw a notice.
+	 *
+	 * @ticket 43056
+	 */
+	public function test_redirect_guess_404_permalink_post_types() {
+		// String format post type.
+		$this->assertCanonical( '/?name=sample-pag&post_type=page', '/sample-page/' );
```
[the remark this answers is not in the document]
Comment (peterwilsoncc) [reply tests/phpunit/tests/canonical.php:285]: > These tests fail for me: phpunit --filter Tests_Canonical

Thanks, I'll check this out.

> There is a typo in the URL parameter: ?name=sample-pag --> ?name=sample-page

That's intended, `sample-pag` will trigger the 404 guessing, I will add a comment to the effect so it doesn't get the spelling fixed in error.

[the remark this answers is not in the document]
Comment (peterwilsoncc) [reply tests/phpunit/tests/canonical.php:291]: I'm not seeing the tests failing with `phpunit --filter Tests_Canonical` running PHP 7.3. Are you using multisite or a different version of PHP?

```
$ phpunit --filter Tests_Canonical
Installing...
Running as single site... To run multisite, use -c tests/phpunit/multisite.xml
Not running ajax tests. To execute these, use --group ajax.
Not running ms-files tests. To execute these, use --group ms-files.
Not running external-http tests. To execute these, use --group external-http.
PHPUnit 7.5.20 by Sebastian Bergmann and contributors.

...............................................................  63 / 143 ( 44%)
............................................................... 126 / 143 ( 88%)
.................                                               143 / 143 (100%)

Time: 15.38 seconds, Memory: 176.50 MB
```

Comment added in 529d11b668cd552d78e421544957f36ca3852511, note unintentional typo in commit message 🙃

File: src/wp-includes/canonical.php [hunk 2]
```diff
@@ -897,7 +897,12 @@ function redirect_guess_404_permalink() {
 
 		// If any of post_type, year, monthnum, or day are set, use them to refine the query.
 		if ( get_query_var( 'post_type' ) ) {
-			$where .= $wpdb->prepare( ' AND post_type = %s', get_query_var( 'post_type' ) );
+			if ( is_array( get_query_var( 'post_type' ) ) ) {
+				// phpcs:ignore WordPress.DB.PreparedSQLPlaceholders.UnfinishedPrepare
+				$where .= " AND post_type IN ('" . join( "', '", array_map( 'esc_sql', get_query_var( 'post_type' ) ) ) . "')";
```
[the remark this answers is not in the document]
Comment (peterwilsoncc) [reply src/wp-includes/canonical.php]: Good catch, removed in 2c826481eeeab65c9f2ba48d7cfe3bf19cf382d1

**Comment by dream-encode** [comment]
Merged into WP Core in https://core.trac.wordpress.org/changeset/49200

**Pull request closed without merging**